composer/callbacks/speed_monitor.py

```diff
@@ -63,106 +63,106 @@ class SpeedMonitor(Callback):
-        self.wall_clock_total = 0.0
+        # Track the epoch num samples and wct to compute throughput over the entire epoch
+        self.epoch_start_num_samples = Time.from_sample(0)
+        self.epoch_start_wct = datetime.timedelta(0)
+
+        # Track the batch num samples and wct to compute throughput over a window of batches
+        self.batch_start_num_samples = Time.from_sample(0)
+        self.batch_start_wct = datetime.timedelta(0)
+        self.rolling_batch_times_and_num_samples: Deque[Tuple[datetime.timedelta,
+                                                              Time[int]]] = deque(maxlen=window_size)
+        self.window_size = window_size
 
-        # log the total wall clock_time that has been spent in training
-        self.wall_clock_train = 0.0
+        # To optimize performance for the the batch wct calculations, keep a running sum over the entire deque
+        self.running_wct = datetime.timedelta(0)
+        self.running_num_samples = Time.from_sample(0)
 
-        # log the total wall clock_time that has been spent in validation
-        self.wall_clock_val = 0.0
+        # Keep track of time spent evaluating
+        self.total_eval_wct = datetime.timedelta(0)
 
-        self.epoch_start_time = 0.0
-        self.validation_start_time = 0.0
-        self.epoch_time_in_validation = 0.0
-        self.batch_start_num_samples = None
-        self.batch_end_times: Deque[float] = deque(maxlen=window_size + 1)  # rolling list of batch end times
-        self.batch_num_samples: Deque[int] = deque(maxlen=window_size)  # rolling list of num samples in batch.
-        self.window_size = window_size
-        self.loaded_state: Optional[Dict[str, Any]] = None
+        self._loaded_state: Optional[Dict[str, Any]] = None
 
     def state_dict(self) -> Dict[str, Any]:
-        current_time = time.time()
         return {
-            "train_examples_per_epoch": self.train_examples_per_epoch,
-            "wall_clock/train": self.wall_clock_train,
-            "wall_clock/val": self.wall_clock_val,
-            "wall_clock/total": self.wall_clock_total,
-            "epoch_duration": current_time - self.epoch_start_time,
-            "batch_durations": [current_time - x for x in self.batch_end_times],
-            "batch_num_samples": self.batch_num_samples,
+            "epoch_start_num_samples": self.epoch_start_num_samples,
+            "epoch_start_wct": self.epoch_start_wct,
+            "batch_start_num_samples": self.batch_start_num_samples,
+            "batch_start_wct": self.batch_start_wct,
+            "rolling_batch_times_and_num_samples": self.rolling_batch_times_and_num_samples,
+            "running_wct": self.running_wct,
+            "running_num_samples": self.running_num_samples,
+            "total_eval_wct": self.total_eval_wct,
         }
 
     def load_state_dict(self, state: Dict[str, Any]) -> None:
-        self.loaded_state = state
+        self._loaded_state = state
```

Comment:
could you add a comment to explain why we don't just load the state direclty here, but instead wait for one of the events?

Reply:
Good question, I think this was a relic from the previous version where we stored the timestamp, rather than the duration, in the internal state. Now this is not needed anymore.